File: excelium-common/src/main/java/excelium/common/EnvironmentUtil.java
/*
 * MIT License
 *
 * Copyright (c) 2018 Excelium
 *
 * Permission is hereby granted, free of charge, to any person obtaining a copy
 * of this software and associated documentation files (the "Software"), to deal
 * in the Software without restriction, including without limitation the rights
 * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
 * copies of the Software, and to permit persons to whom the Software is
 * furnished to do so, subject to the following conditions:
 *
 * The above copyright notice and this permission notice shall be included in all
 * copies or substantial portions of the Software.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
 * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
 * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
 * SOFTWARE.
 */

package excelium.common;

import excelium.model.enums.Browser;
import excelium.model.enums.Platform;
import excelium.model.test.config.Environment;
import excelium.model.test.config.MobileAppEnvironment;
import excelium.model.test.config.MobileWebEnvironment;
import excelium.model.test.config.PcEnvironment;
import org.apache.commons.collections4.CollectionUtils;
import org.apache.commons.lang3.StringUtils;

import java.util.ArrayList;
import java.util.List;

/**
 * Collection of environment handling utilities
 *
 * @author PhungDucKien
 * @since 2018.04.17
 */
public class EnvironmentUtil {

    /**
     * Gets available pc environments.
     *
     * @param browser     the browser
     * @param versions    the versions
     * @param resolutions the resolutions
     * @param baseUrl     the base url
     * @return the available pc environments
     */
    public static List<Environment> getAvailablePcEnvironments(Browser browser, List<String> versions, List<String> resolutions, String baseUrl) {
        List<Environment> environments = new ArrayList<>();
        Platform platform = PlatformDetector.getPlatform();
        if (isBrowserAvailable(platform, browser)) {
            if (CollectionUtils.isEmpty(versions)) {
                versions = new ArrayList<>();
                versions.add("");
            }
            if (CollectionUtils.isEmpty(resolutions)) {
                resolutions = new ArrayList<>();
                resolutions.add("1024x768");
            }
            for (String version : versions) {
                for (String resolution : resolutions) {
                    PcEnvironment environment = new PcEnvironment();
                    environment.setPlatform(platform);
                    environment.setBrowser(browser);
                    environment.setBrowserVersion(version);
                    environment.setResolution(resolution);
                    environment.setBaseUrl(baseUrl);
                    environments.add(environment);
                }
            }
        }
        return environments;
    }

    /**
     * Gets available mobile web environments.
     *
     * @param platform         the platform
     * @param platformVersions the platform versions
     * @param deviceNames      the device names
     * @param orientations     the orientations
     * @param udid             the udid
     * @param browsers         the browsers
     * @param baseUrl          the base url
     * @return the available mobile web environments
     */
    public static List<Environment> getAvailableMobileWebEnvironments(Platform platform, List<String> platformVersions, List<String> deviceNames, List<String> orientations, String udid, List<String> browsers, String baseUrl) {
        if (CollectionUtils.isEmpty(platformVersions)) {
            platformVersions = new ArrayList<>();
            platformVersions.add("");
        }
        if (CollectionUtils.isEmpty(deviceNames)) {
            deviceNames = new ArrayList<>();
            if (platform == Platform.ANDROID) {
                deviceNames.add("Android");
            } else if (platform == Platform.IOS) {
                deviceNames.add("iPhone");
            }
        }
        if (CollectionUtils.isEmpty(orientations)) {
            orientations = new ArrayList<>();
            orientations.add("Portrait");
        }
        if (CollectionUtils.isEmpty(browsers)) {
            browsers = new ArrayList<>();
            if (platform == Platform.ANDROID) {
                browsers.add("Chrome");
            } else if (platform == Platform.IOS) {
                browsers.add("Safari");
            }
        }
        List<Environment> environments = new ArrayList<>();
        for (String platformVersion : platformVersions) {
            for (String deviceName : deviceNames) {
                for (String orientation : orientations) {
                    for (String browser : browsers) {
                        MobileWebEnvironment environment = new MobileWebEnvironment();
                        environment.setPlatform(platform);
                        environment.setPlatformVersion(platformVersion);
                        environment.setDeviceName(deviceName);
                        environment.setOrientation(orientation);
                        environment.setUdid(StringUtils.isBlank(udid) ? "auto" : udid);
                        environment.setBrowser(Browser.fromName(browser));
                        environment.setBaseUrl(baseUrl);
                        environments.add(environment);
                    }
                }
            }
        }
        return environments;
    }

    /**
     * Gets available mobile app environments.
     *
     * @param platform         the platform
     * @param platformVersions the platform versions
     * @param deviceNames      the device names
     * @param orientations     the orientations
     * @param udid             the udid
     * @param appPath          the app path
     * @param appActivity      the app activity
     * @param appPackage       the app package
     * @param appWaitActivity  the app wait activity
     * @param appWaitPackage   the app wait package
     * @param bundleId         the bundle ID
     * @return the available mobile app environments
     */
    public static List<Environment> getAvailableMobileAppEnvironments(Platform platform, List<String> platformVersions, List<String> deviceNames, List<String> orientations, String udid,
                                                                      String appPath, String appActivity, String appPackage, String appWaitActivity, String appWaitPackage, String bundleId) {
        if (CollectionUtils.isEmpty(platformVersions)) {
            platformVersions = new ArrayList<>();
            platformVersions.add("");
        }
        if (CollectionUtils.isEmpty(deviceNames)) {
            deviceNames = new ArrayList<>();
            if (platform == Platform.ANDROID) {
                deviceNames.add("Android");
            } else if (platform == Platform.IOS) {
                deviceNames.add("iPhone");
            }
        }
        if (CollectionUtils.isEmpty(orientations)) {
            orientations = new ArrayList<>();
            orientations.add("Portrait");
        }
        List<Environment> environments = new ArrayList<>();
        for (String platformVersion : platformVersions) {
            for (String deviceName : deviceNames) {
                for (String orientation : orientations) {
                    MobileAppEnvironment environment = new MobileAppEnvironment();
                    environment.setPlatform(platform);
                    environment.setPlatformVersion(platformVersion);
                    environment.setDeviceName(deviceName);
                    environment.setOrientation(orientation);
                    environment.setUdid(StringUtils.isBlank(udid) ? "auto" : udid);
                    environment.setAppPath(appPath);
                    environment.setAppActivity(appActivity);
                    environment.setAppPackage(appPackage);
                    environment.setAppWaitActivity(appWaitActivity);
                    environment.setAppWaitPackage(appWaitPackage);
                    environment.setBundleId(bundleId);
                    environments.add(environment);
                }
            }
        }
        return environments;
    }

    /**
     * Checks if the given browser is available on the platform
     *
     * @param platform the platform
     * @param browser  the browser
     * @return is the browser available
     */
    private static boolean isBrowserAvailable(Platform platform, Browser browser) {
        switch (browser) {
            case IE:
            case EDGE:
                return platform.isWindows();
            case SAFARI:
                return platform.isMac();
            default:
                return true;
        }
    }
}

File: packages/common/src/EnvironmentUtil.ts
/*
 * Licensed to the Software Freedom Conservancy (SFC) under one
 * or more contributor license agreements.  See the NOTICE file
 * distributed with this work for additional information
 * regarding copyright ownership.  The SFC licenses this file
 * to you under the Apache License, Version 2.0 (the
 * "License"); you may not use this file except in compliance
 * with the License.  You may obtain a copy of the License at
 *
 *   http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing,
 * software distributed under the License is distributed on an
 * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
 * KIND, either express or implied.  See the License for the
 * specific language governing permissions and limitations
 * under the License.
 */

import { Browser, Environment, MobileAppEnvironment, MobileWebEnvironment, PcEnvironment, Platform } from '@excelium/model';
import CollectionUtil from './CollectionUtil';
import PlatformDetector from './PlatformDetector';
import StringUtil from './StringUtil';

/**
 * Collection of environment handling utilities
 *
 * @author PhungDucKien
 * @since 2018.04.17
 */
export default class EnvironmentUtil {
  /**
   * Gets available pc environments.
   *
   * @param browser     the browser
   * @param versions    the versions
   * @param resolutions the resolutions
   * @param baseUrl     the base url
   * @return the available pc environments
   */
  public static getAvailablePcEnvironments(browser: Browser, versions: string[], resolutions: string[], baseUrl: string | null): Environment[] {
    const environments = [] as Environment[];
    const platform = PlatformDetector.getPlatform();
    if (this.isBrowserAvailable(platform, browser)) {
      if (CollectionUtil.isEmpty(versions)) {
        versions = [''];
      }
      if (CollectionUtil.isEmpty(resolutions)) {
        resolutions = ['1024x768'];
      }
      for (const version of versions) {
        for (const resolution of resolutions) {
          const environment = new PcEnvironment();
          environment.platform = platform;
          environment.browser = browser;
          environment.browserVersion = version;
          environment.resolution = resolution;
          environment.baseUrl = baseUrl;
          environments.push(environment);
        }
      }
    }
    return environments;
  }

  /**
   * Gets available mobile web environments.
   *
   * @param platform         the platform
   * @param platformVersions the platform versions
   * @param deviceNames      the device names
   * @param orientations     the orientations
   * @param udid             the udid
   * @param browsers         the browsers
   * @param baseUrl          the base url
   * @return the available mobile web environments
   */
  public static getAvailableMobileWebEnvironments(
    platform: Platform,
    platformVersions: string[],
    deviceNames: string[],
    orientations: string[],
    udid: string | null,
    browsers: string[],
    baseUrl: string | null
  ): Environment[] {
    if (CollectionUtil.isEmpty(platformVersions)) {
      platformVersions = [''];
    }
    if (CollectionUtil.isEmpty(deviceNames)) {
      deviceNames = [];
      if (platform === Platform.ANDROID) {
        deviceNames.push('Android');
      } else if (platform === Platform.IOS) {
        deviceNames.push('iPhone');
      }
    }
    if (CollectionUtil.isEmpty(orientations)) {
      orientations = ['Portrait'];
    }
    if (CollectionUtil.isEmpty(browsers)) {
      browsers = [];
      if (platform === Platform.ANDROID) {
        browsers.push('Chrome');
      } else if (platform === Platform.IOS) {
        browsers.push('Safari');
      }
    }
    const environments = [] as Environment[];
    for (const platformVersion of platformVersions) {
      for (const deviceName of deviceNames) {
        for (const orientation of orientations) {
          for (const browser of browsers) {
            const environment = new MobileWebEnvironment();
            environment.platform = platform;
            environment.platformVersion = platformVersion;
            environment.deviceName = deviceName;
            environment.orientation = orientation;
            environment.udid = StringUtil.isBlank(udid) ? 'auto' : udid;
            environment.browser = Browser.fromName(browser);
            environment.baseUrl = baseUrl;
            environments.push(environment);
          }
        }
      }
    }
    return environments;
  }

  /**
   * Gets available mobile app environments.
   *
   * @param platform         the platform
   * @param platformVersions the platform versions
   * @param deviceNames      the device names
   * @param orientations     the orientations
   * @param udid             the udid
   * @param appPath          the app path
   * @param appActivity      the app activity
   * @param appPackage       the app package
   * @param appWaitActivity  the app wait activity
   * @param appWaitPackage   the app wait package
   * @param bundleId         the bundle ID
   * @return the available mobile app environments
   */
  public static getAvailableMobileAppEnvironments(
    platform: Platform,
    platformVersions: string[],
    deviceNames: string[],
    orientations: string[],
    udid: string | null,
    appPath: string | null,
    appActivity: string | null,
    appPackage: string | null,
    appWaitActivity: string | null,
    appWaitPackage: string | null,
    bundleId: string | null
  ): Environment[] {
    if (CollectionUtil.isEmpty(platformVersions)) {
      platformVersions = [''];
    }
    if (CollectionUtil.isEmpty(deviceNames)) {
      deviceNames = [];
      if (platform === Platform.ANDROID) {
        deviceNames.push('Android');
      } else if (platform === Platform.IOS) {
        deviceNames.push('iPhone');
      }
    }
    if (CollectionUtil.isEmpty(orientations)) {
      orientations = ['Portrait'];
    }
    const environments = [] as Environment[];
    for (const platformVersion of platformVersions) {
      for (const deviceName of deviceNames) {
        for (const orientation of orientations) {
          const environment = new MobileAppEnvironment();
          environment.platform = platform;
          environment.platformVersion = platformVersion;
          environment.deviceName = deviceName;
          environment.orientation = orientation;
          environment.udid = StringUtil.isBlank(udid) ? 'auto' : udid;
          environment.appPath = appPath;
          environment.appActivity = appActivity;
          environment.appPackage = appPackage;
          environment.appWaitActivity = appWaitActivity;
          environment.appWaitPackage = appWaitPackage;
          environment.bundleId = bundleId;
          environments.push(environment);
        }
      }
    }
    return environments;
  }

  /**
   * Checks if the given browser is available on the platform
   *
   * @param platform the platform
   * @param browser  the browser
   * @return is the browser available
   */
  private static isBrowserAvailable(platform: Platform, browser: Browser): boolean {
    switch (browser) {
      case Browser.IE:
      case Browser.EDGE:
        return Platform.isWindows(platform);
      case Browser.SAFARI:
        return Platform.isMac(platform);
      default:
        return true;
    }
  }
}
